File: locals/solutions/alex/alexTimothy.c
#include <math.h>
#include <stdio.h>

#define EPS 1e-6
#define RADIUS 6371
#define PI 3.14159265358979

typedef struct Vec {
	double x, y, z;
} Vec;

double min(double a, double b) {
	return a < b ? a : b;
}

Vec newVec(double x, double y, double z) {
	Vec v;
	v.x = x;
	v.y = y;
	v.z = z;
	return v;
}

Vec vecAdd(Vec a, Vec b) {
	return newVec(a.x + b.x, a.y + b.y, a.z + b.z);
}

Vec vecSub(Vec a, Vec b) {
	return newVec(a.x - b.x, a.y - b.y, a.z - b.z);
}

double vecDot(Vec a, Vec b) {
	return a.x * b.x + a.y * b.y + a.z * b.z;
}

double vecMag2(Vec v) {
	return vecDot(v, v);
}

double vecMag(Vec v) {
	return sqrt(vecMag2(v));
}

Vec vecScale(Vec v, double s) {
	return newVec(v.x * s, v.y * s, v.z * s);
}

Vec vecScaleTo(Vec v, double s) {
	return vecScale(v, s / vecMag(v));
}
		
Vec vecNormalize(Vec v) {
	return vecScaleTo(v, 1);
}

Vec vecCross(Vec a, Vec b) {
	return newVec(a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x);
}

// Given a plane defined by a, b, and c, returns the projection of p onto the plane.
Vec vecProj(Vec a, Vec b, Vec c, Vec p) {
	Vec ab = vecSub(b, a);
	Vec ac = vecSub(c, a);
	Vec norm = vecNormalize(vecCross(ab, ac));
	return vecSub(p, vecScale(norm, vecDot(vecSub(p, a), norm)));
}

int sig(double x) {
	if (fabs(x) < EPS)
		return 0;
	return x < 0 ? -1 : 1;
}
	
// Given the latitude and longitude (in degrees), returns the 3d point
// on the sphere of radius 6371km centered around the origin.
Vec getPointOnEarth(double lat, double lon) {
	lat = lat * PI / 180;
	lon = lon * PI / 180;
	double x = cos(lat) * cos(lon);
	double y = cos(lat) * sin(lon);
	double z = sin(lat);
	return vecScale(newVec(x, y, z), RADIUS);
}
	
int main(void) {
	int n;
	// Read in number of trips.
	scanf("%d", &n);
	
	// Get the points for the northPole and the origin.
	Vec northPole = getPointOnEarth(90, 0);
	Vec origin = newVec(0, 0, 0);
	for (int tripI = 0; tripI < n; tripI++) {
		double lat0, lon0, lat1, lon1;
		scanf("%lf", &lat0);
		scanf("%lf", &lon0);
		scanf("%lf", &lat1);
		scanf("%lf", &lon1);

		int onEquator = fabs(lat0) < 1e-9 && fabs(lat1) < 1e-9;
		
		// Get the points of the two cities.
		Vec p0 = getPointOnEarth(lat0, lon0);
		Vec p1 = getPointOnEarth(lat1, lon1);

		// Calculate the initial minimum distance.
		double minDist = min(vecMag(vecSub(p0, northPole)), vecMag(vecSub(p1, northPole)));
		
		// Get a vector perpendicular to the plane defined by the origin and the two cities.
		Vec norm = vecCross(p0, p1);
		// Project the north pole on to the plane defined by the origin and the two cities.
		Vec proj = vecProj(origin, p0, p1, northPole);
		
		// Get the sign of the cross product (on the plane) from the projected north pole to both cities.
		int sig0 = sig(vecDot(vecCross(proj, p0), norm));
		int sig1 = sig(vecDot(vecCross(proj, p1), norm));

		// If the projected point is between the smallest arc formed by the two, Alex is right.
		if (!onEquator && sig0 == -1 && sig1 == 1) {
			double originalMinDist = minDist;
			// Get the true minimum distance.
			minDist = vecMag(vecSub(northPole, vecScaleTo(proj, RADIUS)));
			printf("Alex\n");
		} else {
			printf("Timothy\n");
		}
		// Print out the minimum distance.
		printf("%.6f\n\n", minDist);
	}
	return 0;
}

File: locals/solutions/alex/alexTimothy.java
import java.util.*;
import java.io.*;

/*

2
38.946474 -77.437568
39.918620 116.443983

28.430849 -81.324178
51.507370 -0.148300


*/


public class alexTimothy {

	static double RADIUS = 6371;
	static double eps = 1e-6;
	
	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		// Read in number of test cases.
		int T = in.nextInt();
		
		Vec northPole = getPointOnEarth(90, 0);
		Vec center = new Vec(0, 0, 0);
		
		for (int tripI = 1; tripI <= T; tripI++) {
			// Read in the coordinates of the departure city.
			double lat0 = in.nextDouble();
			double lon0 = in.nextDouble();

			// Read in the coordinates of the destination  city.
			double lat1 = in.nextDouble();
			double lon1 = in.nextDouble();
			
			validateInput(tripI, lat0, lon0, lat1, lon1);
			
			// Get the 3D points for the departure and destination cities.
			Vec p0 = getPointOnEarth(lat0, lon0);
			Vec p1 = getPointOnEarth(lat1, lon1);
			
			if (p0.sub(p1).mag() < eps) {
				exitWithDataValidationError("Case " + tripI + ": Departure and destination cities too close!");
			}
			if (p0.sub(northPole).mag() < eps) {
				exitWithDataValidationError("Case " + tripI + ": Departure city too close to North Pole!");
			}
			if (p1.sub(northPole).mag() < eps) {
				exitWithDataValidationError("Case " + tripI + ": Destination city too close to North Pole!");
			}
			if (p0.sub(center).cross(p1.sub(center)).mag() < eps) {
				exitWithDataValidationError("Case " + tripI + ": Only one shortest path should exist!");
			}
			
			// Initialize the answer (minimum distance) as the min of the
			// distance from the Noth Pole to both the departure and destination cities.
			double ans = Math.min(p0.sub(northPole).mag(), p1.sub(northPole).mag());
			
			// Project the north pole onto the plane defined by the center and the
			// two cities.
			Vec proj = Vec.proj(center, p0, p1, northPole);

			// Start by assuming that Timothy is right.
			String personRight = "Timothy";
			
			Vec projXY = moveToXYPlane(center, p0, p1, proj);
			Vec northPoleXY = moveToXYPlane(center, p0, p1, northPole);
			Vec p0XY = moveToXYPlane(center, p0, p1, p0);
			Vec p1XY = moveToXYPlane(center, p0, p1, p1);
			Vec centerXY = moveToXYPlane(center, p0, p1, center);
			
			Vec ray = projXY.sub(centerXY);
			
			// If the flight path is not just along the equator,
			// there is a chance that Alex is right.
			if (ray.mag2() > eps) {
				// Calculate the minimum distance from the North Pole to a point on the
				// great circle defined by the two cities.
				double dxy = RADIUS - proj.sub(center).mag();
				double dz2 = northPole.sub(proj).mag2();
				double dist = Math.sqrt(dxy * dxy + dz2);
				
				Vec ray0 = p0XY.sub(centerXY);
				Vec ray1 = p1XY.sub(centerXY);
				double cross0 = ray0.cross(ray).z;
				double cross1 = ray.cross(ray1).z;
				double cross01 = ray0.cross(ray1).z;
				
				int s0 = (int) Math.signum(cross0);
				int s1 = (int) Math.signum(cross1);
				int s01 = (int) Math.signum(cross01);
				
				if ((s0 == s01 && s1 == s01) || Math.abs(cross0) < 1e-6 || Math.abs(cross1) < 1e-6) {
					personRight = "Alex";
					if (Math.abs(ans - dist) < 1.001) {
						exitWithDataValidationError("Case " + tripI + ": Answer too close");
					}
					ans = dist;
				}
			}
			
			System.out.println(personRight);
			System.out.printf("%.6f\n\n", ans);
		}
	}
	
	// Given 3 non-colinear points that define a plane and a point on that plane,
	// Project that point to the XY-plane retaining relative positioning and scale on the plane.
	static Vec moveToXYPlane(Vec a, Vec b, Vec c, Vec p) {
		Vec vx = b.sub(a).normalize();
		Vec vy = c.sub(a).sub(vx.scale(vx.dot(c.sub(a)))).normalize();
		return new Vec(vx.dot(p), vy.dot(p), 0);
	}
	
	// Given the latitude and longitude (in degrees), returns the 3d point
	// on the sphere of radius 6371km centered around the origin.
	static Vec getPointOnEarth(double lat, double lon) {
		lat = lat * Math.PI / 180;
		lon = lon * Math.PI / 180;
		double x = Math.cos(lat) * Math.cos(lon);
		double y = Math.cos(lat) * Math.sin(lon);
		double z = Math.sin(lat);
		return new Vec(x, y, z).scale(RADIUS);
	}
	
	static void validateInput(int tripI, double lat0, double lon0, double lat1, double lon1) {
		if (lat0 < -90 || lat0 > 90) {
			exitWithDataValidationError("Case " + tripI + ": Departure city latitutde " + lat0 + " out of range!");
		}
		if (lat1 < -90 || lat1 > 90) {
			exitWithDataValidationError("Case " + tripI + ": Destination city latitutde " + lat1 + " out of range!");
		}
		if (lon0 < -180 || lon0 > 180) {
			exitWithDataValidationError("Case " + tripI + ": Departure city longitude " + lon0 + " out of range!");
		}
		if (lon1 < -180 || lon1 > 180) {
			exitWithDataValidationError("Case " + tripI + ": Destination city longitude " + lon1 + " out of range!");
		}
	}
	
	static void exitWithDataValidationError(String errorMessage) {
		System.err.println("Data validation error:");
		System.err.println(errorMessage);
		System.exit(1);
	}
	
	static class Vec {
		double x, y, z;
		
		public Vec(double xx, double yy, double zz) {
			x = xx;
			y = yy;
			z = zz;
		}
		
		public Vec add(Vec v) {
			return new Vec(x + v.x, y + v.y, z + v.z);
		}
		
		public Vec sub(Vec v) {
			return new Vec(x - v.x, y - v.y, z - v.z);
		}
		
		public Vec scale(double s) {
			return new Vec(x * s, y * s, z * s);
		}
		
		public Vec scaleTo(double s) {
			return scale(s / mag());
		}
		
		public Vec normalize() {
			return scaleTo(1);
		}
		
		public Vec cross(Vec v) {
			return new Vec(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x);
		}
		
		public double dot(Vec v) {
			return x * v.x + y * v.y + z * v.z;
		}
		
		public double mag2() {
			return dot(this);
		}
		
		public double mag() {
			return Math.sqrt(mag2());
		}
		
		// Renters the minimum angle between this vector and the vector v.
		public double angleBetween(Vec v) {
			return Math.acos( Math.max(-1, Math.min(1, dot(v) / (mag() * v.mag()))) );
		}
		
		// Given a plane defined by a, b, and c, returns the projection of p onto the plane.
		public static Vec proj(Vec a, Vec b, Vec c, Vec p) {
			Vec ab = b.sub(a);
			Vec ac = c.sub(a);
			Vec norm = ab.cross(ac).scaleTo(1);
			return p.sub(norm.scale(p.sub(a).dot(norm)));
		}
		
		public String toString() {
			return String.format("<%.3f, %.3f, %.3f>", x, y, z);
		}
	}
}